### PR TITLE
Add random_transform_generator for augmenting images.

### PR DESCRIPTION
**Pull request by de-vri-es**
This PR adds a `random_transform_generator` to be used for augmenting data. It replaces the use of `keras.preprocessing.image.ImageDataGenerator`.

The main reason for doing this is that `ImageDataGenerator` creates a random transformation internally and applies it directly to the image. That doesn't give us the chance to apply the same transformation to bounding boxes.

The current workaround is to seed the global PRNG used by `ImageDataGenerator` with a random value and then have it transform first the image data, and then a mask for the bounding box (with the PRNG reset to the same random seed). The new bounding box is then determined from the transformed mask.

This is a rather expensive process, but also slightly fragile to chances in the `ImageDataGenerator`. It relies on the fact that `ImageDataGenerator` uses the global numpy PRNG, and that the number and order of generating random numbers is the same for the original image and the mask.

In contrast, the new `random_transform_generator` doesn't apply any transformations, it just generates them. The responsibility for applying the transforms to the image and bounding box has been moved into our own `Generator` with this PR.

`random_transform_generator` supports the following transformations:
* rotation
* translation
* scaling
* shearing
* flips

Notably, the one thing from `ImageDataGenerator.random_transform` it does not support is channel shifting, since that is not a homogeneous linear transformation in 2D.

### CODE REVIEW

File: keras_retinanet/preprocessing/generator.py [hunk 1]
```diff
@@ -112,19 +116,32 @@ def resize_image(self, image):
...
+        if self.transform_generator is not None:
+            transform = adjust_transform_for_image(next(self.transform_generator), image)
+            apply_transform(transform, image, self.transform_parameters)
 
-            # randomly transform both image and annotations
-            image, annotations = random_transform(image, annotations, self.image_data_generator)
+            # Transform the bounding boxes in the annotations.
+            annotations = annotations.copy()
+            for index in range(annotations.shape[0]):
+                annotations[index, :4] = transform_aabb(transform, annotations[index, :4])
```
Comment (de-vri-es) [inline keras_retinanet/preprocessing/generator.py]: One thing I was wondering: What should the generator do if `transform_aabb` puts the annotation outside of the image canvas? I'm assuming they will automatically be filtered now by `compute_input_output`. But maybe I overlooked something?

**Comment by de-vri-es** [comment]
Note: should fix #68, #181 and I think #186.

**Comment by de-vri-es** [comment]
Note: due to the holiday period, reviews are going to be slightly delayed. That also goes for this PR.

File: keras_retinanet/preprocessing/generator.py (hunk 1 above)
Comment (de-vri-es) [reply keras_retinanet/preprocessing/generator.py]: Also, is that desirable? When the aabb ends up outside the image and it is filtered, that means that part of the image becomes background, even though it was actually an annotation.

Should we instead discard the entire image? Or should we try up to X times to get a transformation that doesn't put any annotations outside of the image (and discard if that failed X times)? Or should we clip the new bounding box with the image canvas?

File: keras_retinanet/preprocessing/generator.py [hunk 2]
```diff
@@ -112,19 +116,32 @@ def resize_image(self, image):
...
-    def preprocess_group(self, image_group, annotations_group):
-        for index, (image, annotations) in enumerate(zip(image_group, annotations_group)):
-            # preprocess the image (subtract imagenet mean)
-            image = self.preprocess_image(image)
+    def preprocess_group_entry(self, image, annotations):
+        # preprocess the image
+        image = self.preprocess_image(image)
+
+        # randomly transform both image and annotations
+        if self.transform_generator is not None:
```
Comment (hgaiser) [inline keras_retinanet/preprocessing/generator.py]: Why not use `if self.transform_generator:` ?

File: keras_retinanet/preprocessing/generator.py (hunk 1 above)
Comment (hgaiser) [reply keras_retinanet/preprocessing/generator.py]: My first thought is to clip the annotation to the image canvas, to preserve as much of the annotations as possible. What do you think ?

File: keras_retinanet/utils/transform.py [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+    ])
+
+    # Extract the min and max corners again.
+    min_corner = points.min(axis=1)
+    max_corner = points.max(axis=1)
+
+    return [min_corner[0], min_corner[1], max_corner[0], max_corner[1]]
+
+
+def _random_vector(min, max, prng = DEFAULT_PRNG):
```
Comment (hgaiser) [inline keras_retinanet/utils/transform.py]: Non-aligned default arguments are written down without spaces around the equal sign (same for some other places).

File: keras_retinanet/utils/transform.py [hunk 4]
```diff
@@ -0,0 +1,241 @@
...
+        the translation matrix as 3 by 3 numpy array
+    """
+    return np.array([
+        [1, 0, translation[0]],
+        [0, 1, translation[1]],
+        [0, 0, 1]
+    ])
+
+
+def random_translation(min, max, prng = DEFAULT_PRNG):
```
Comment (hgaiser) [inline keras_retinanet/utils/transform.py]: Do we want to distinguish between X/Y translation min/max?

File: keras_retinanet/utils/transform.py [hunk 5]
```diff
@@ -0,0 +1,241 @@
...
+    # Returns
+        a homogeneous 3 by 3 translation matrix
+    """
+    return translation(_random_vector(min, max, prng))
+
+
+def shear(amount):
+    """ Construct a homogeneous 2D shear matrix.
+    # Arguments
+        amount: the shear amount
```
Comment (hgaiser) [inline keras_retinanet/utils/transform.py]: Maybe make it a bit more clear what it is. Presumably it is the angle of the shear in radians?

File: keras_retinanet/utils/transform.py [hunk 6]
```diff
@@ -0,0 +1,241 @@
...
+        [1, -np.sin(amount), 0],
+        [0,  np.cos(amount), 0],
+        [0, 0, 1]
+    ])
+
+
+def random_shear(min, max, prng = DEFAULT_PRNG):
+    """ Construct a random 2D shear matrix with shear angle between -max and max.
+    # Arguments
+        min:  the minumum shear factor.
```
Comment (hgaiser) [inline keras_retinanet/utils/transform.py]: Same here.

File: keras_retinanet/utils/transform.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+        the zoom matrix as 3 by 3 numpy array
+    """
+    return np.array([
+        [factor[0], 0, 0],
+        [0, factor[1], 0],
+        [0, 0, 1]
+    ])
+
+
+def random_scaling(min, max, prng = DEFAULT_PRNG):
```
Comment (hgaiser) [inline keras_retinanet/utils/transform.py]: Same here, do we want to distinguish between x/y ?

File: keras_retinanet/utils/transform.py (hunk 3 above)
Comment (philvdm) [reply keras_retinanet/utils/transform.py]: pep8 could be used for catching those spaces

Comment (de-vri-es) [reply keras_retinanet/utils/transform.py]: One problem with using `pytest --pep8` for that is that we do use spaces around the parameter assignment on multi-line function calls for alignment. The pep8 test doesn't distinguish between that and this type of whitespace.

File: keras_retinanet/utils/transform.py (hunk 4 above)
Comment (de-vri-es) [reply keras_retinanet/utils/transform.py]: At this point I would say yes. These are generic transform functions. My initial reasoning was that you might want to scale the max and min with the image width and height. That isn't necessary now since the data generator does the scaling itself after generating the transform.

Still, I think it's a valid use case for a generic transform function.

File: keras_retinanet/utils/transform.py (hunk 5 above)
Comment (de-vri-es) [reply keras_retinanet/utils/transform.py]: It is an angle, though the matrix generated isn't actually a shear matrix. However, I copied it from `keras` since I figured we should probably use the same definition of shear as `keras` does, even if it is technically incorrect.

I believe the `keras` definition of shear compensates for the stretching that a real shear operation would do by reducing the Y component as the X component gets larger. So it's sort of halfway between a shear and a rotation.

Another thing I wasn't really happy about is that this shear functions only shears parallel to the X axis. It's not very generic.

File: keras_retinanet/preprocessing/generator.py (hunk 2 above)
Comment (de-vri-es) [reply keras_retinanet/preprocessing/generator.py]: No reason really. Do you prefer it like that?

File: keras_retinanet/preprocessing/generator.py (hunk 1 above)
Comment (de-vri-es) [reply keras_retinanet/preprocessing/generator.py]: I'm not sure. I have a slight feeling that it might reduce the quality of training in some cases. Maybe some very distinctive part of the thing to recognize is clipped off, or maybe the clipped thing looks more like a different object (contrived example: a face with a clown nose vs a face with a normal nose).

File: keras_retinanet/utils/transform.py (hunk 3 above)
Comment (de-vri-es) [reply keras_retinanet/utils/transform.py]: Anyway, fixed, I think.

File: keras_retinanet/utils/transform.py (hunk 4 above)
Comment (de-vri-es) [reply keras_retinanet/utils/transform.py]: Damn you github, this comment thread is not outdated -.-

File: keras_retinanet/preprocessing/generator.py (hunk 2 above)
Comment (hgaiser) [reply keras_retinanet/preprocessing/generator.py]: Yeah.

File: keras_retinanet/utils/transform.py (hunk 4 above)
Comment (hgaiser) [reply keras_retinanet/utils/transform.py]: But this code handles X and Y the same way right? Using the same min/max?

Comment (de-vri-es) [reply keras_retinanet/utils/transform.py]: no, min and max are 2D vectors (or a list/tuple/sequence of two elements)

Comment (hgaiser) [reply keras_retinanet/utils/transform.py]: Ah I see, I missed that part then. :+1:

File: keras_retinanet/preprocessing/generator.py (hunk 1 above)
Comment (hgaiser) [reply keras_retinanet/preprocessing/generator.py]: I think filtering it out entirely is more harmful then clipping. So either we make sure it doesn't occur, or we clip it in my opinion.

**Comment by hgaiser** [comment]
As far as I can tell we only need to decide on what to do when an annotation is outside of the image canvas after transformation. When does this happen? Shouldn't the annotation always remain in the image canvas (aside from some rounding errors perhaps)?

**Comment by de-vri-es** [comment]
No, rotations can easily put annotations completely or partly out of the canvas. The same is true for shearing and translation. Only flips don't have this problem, really.

**Comment by yhenon** [comment]
Two things:
- Currently bounding boxes that overlap an anchor with 0.4 < IoU < 0.5 are tagges as `don't care` and do not contribute to the loss. It might make sense to consider doing something similar for boxes that are clipped by a transformation.
- whatever the ultimate decision, we should let the threshold be user configurable.

File: keras_retinanet/preprocessing/generator.py (hunk 2 above)
Comment (de-vri-es) [reply keras_retinanet/preprocessing/generator.py]: Updated.

**Comment by hgaiser** [comment]
Shall we move on to merging this, and worry about that edge case in a later PR?

**Comment by LaurensHagendoorn** [comment]
If by the edge case you mean when bounding boxes end up outside the image, I think it should be fixed first. I have been getting that result with the current augmentations quite a lot, and it is a problem. I suppose clipping when possible, and not crashing otherwise would be my preferred solution.

**Comment by hgaiser** [comment]
Crashing is not supposed to happen. I propose to indeed either clip for now, or leave it untouched. Personally I still believe clipping would be fine.

**Comment by LaurensHagendoorn** [comment]
Yeah I think crashing was actually fixed by a change a while back. I am in favour of merging this soon though! I have been oogling this branch for a while now, would be awesome to be able to use this!

**Comment by de-vri-es** [comment]
I also don't think it will crash now. I believe it will generate invalid annotations which will be filtered out automatically. For my part, we can make an issue to resolve the discarding/clipping problem and merge as is.

**Comment by de-vri-es** [comment]
I tagged a 0.1 pre-release, since we seemed to be pretty stable.

Rebased this branch on master. Once CI passes I'll merge.

**Comment by de-vri-es** [comment]
Discussion on what to do with the annotations partially transformed out of the canvas can continue at #223
